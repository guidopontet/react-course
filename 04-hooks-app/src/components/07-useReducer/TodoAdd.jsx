import React from 'react'

import useForm from '../../hooks/useForm'

export const TodoAdd = ({ handleAddTodo }) => {
  const [ { description }, handleInputChange, resetForm ] = useForm({
    description: '',
  })

  const handleSubmit = (e) => {
    e.preventDefault();

    if (description.trim().length < 1 ) { return; }

    const newTodo = {
      id: new Date().getTime(),
      description: description,
      done: false,
    };

    const action = {
      type: 'ADD',
      payload: newTodo,
    }

    handleAddTodo(action);
    resetForm();
  }

  return (
    <>
      <h4>Agregar TODO</h4>
      <hr />

      <form onSubmit={ handleSubmit }>
        <input type="text"
                className='form-control'
                name='description'
                placeholder='Ingresa la tarea'
                autoComplete='off'
                value={ description }
                onChange={ handleInputChange }/>

        <button type='submit'
                className='btn btn-primary btn-block mt-3'>
          Agregar
        </button>
      </form>
    </>
  )
}
