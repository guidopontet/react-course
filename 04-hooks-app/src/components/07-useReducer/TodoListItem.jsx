import React from 'react'

export const TodoListItem = ({ todo, index, handleToggle, handleDelete}) => {
  return (
    <li key={ todo.id }
        className='list-group-item'>
      <p className={ todo.done ? 'complete' : '' }
          onClick={ () => handleToggle(todo) }>
        { index + 1 }. { todo.text }
      </p>

      <button className='btn btn-danger'
              onClick={ () => handleDelete(todo) }>Borrar</button>
    </li>
  )
}
