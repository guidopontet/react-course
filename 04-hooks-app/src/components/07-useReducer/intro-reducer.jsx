const initialState = [{
  id: 1,
  todo: "Comprar pan",
  done: false,
}];

const todoReducer = (state = initialState, action) => {
  if (action?.type === "ADD_TODO") {
    return [...state, action.payload];
  }

  return state;
}

let todos = todoReducer();

const newTodo = {
  id: 2,
  todo: "Comprar pasta",
  done: false,
};

const addTodoAction = {
  type: "ADD_TODO",
  payload: newTodo,
}

todos = todoReducer(todos, addTodoAction);

console.log(todos);
