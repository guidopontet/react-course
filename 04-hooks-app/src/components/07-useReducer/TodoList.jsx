import React from 'react'

import { TodoListItem } from './TodoListItem';

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
  return (
    <ul className='list-group list-group-flush'>
      {
        todos.map((todo, i) => {
          return (
            <TodoListItem todo={ todo }
                          index={ i }
                          key={ todo.id }
                          handleDelete= { handleDelete }
                          handleToggle={ handleToggle }/>
          );
        })
      }
    </ul>
  )
}
