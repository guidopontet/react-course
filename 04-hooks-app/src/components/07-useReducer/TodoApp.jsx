import React, { useEffect, useReducer } from 'react';

import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';

import './styles.css';

const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
}

export const TodoApp = () => {
  const [ todos, dispatch ] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleDelete = (todo) => {
    const action = {
      type: 'DELETE',
      payload: todo.id,
    }

    dispatch(action);
  }

  const handleToggle = (todo) => {
    const action = {
      type: 'TOGGLE',
      payload: todo.id,
    }

    dispatch(action);
  }

  const handleAddTodo = ((action) => {
    dispatch(action);
  })

  return (
    <div className='container my-3'>
      <h1>TodoApp ({ todos.length })</h1>
      <hr />

      <div className='row'>
        <div className='col-7'>
          <TodoList todos={ todos }
                    handleToggle={ handleToggle }
                    handleDelete= { handleDelete } />
        </div>
        <div className='col-5'>
          <TodoAdd handleAddTodo={ handleAddTodo }/>
        </div>
      </div>
    </div>
  )
}
