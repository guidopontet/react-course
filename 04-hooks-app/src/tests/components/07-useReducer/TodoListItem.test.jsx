import { shallow } from 'enzyme';

import { TodoListItem } from '../../../components/07-useReducer/TodoListItem';
import { demoTodos } from '../../fixtures/demoTodos';

describe('TodoListItem', () => {
  const handleDelete = jest.fn();
  const handleToggle = jest.fn();

  const wrapper = shallow(
    <TodoListItem
      todo={demoTodos[0]}
      index={0}
      handleDelete={handleDelete}
      handleToggle={handleToggle}
    />
  );

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call handleDelete function', () => {
    wrapper.find('button').simulate('click');
    expect(handleDelete).toHaveBeenCalledWith(demoTodos[0]);
  });

  it('should call handleToggle function', () => {
    wrapper.find('p').simulate('click');
    expect(handleToggle).toHaveBeenCalledWith(demoTodos[0]);
  });

  it('should show text correctly', () => {
    const p = wrapper.find('p');
    expect(p.text()).toBe(`1. ${demoTodos[0].text}`);
  });

  it('should have complete class if item is done', () => {
    const todo = demoTodos[0];
    todo.done = true;

    const wrapper = shallow(<TodoListItem todo={todo} />);

    expect(wrapper.find('p').hasClass('complete')).toBe(true);
  });
});
