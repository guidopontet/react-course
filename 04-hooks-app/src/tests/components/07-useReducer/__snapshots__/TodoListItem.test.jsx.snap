// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem should render correctly 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className=""
    onClick={[Function]}
  >
    1
    . 
    Learn React
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
